contracts/proxies/CurrencyConvertor.sol

```diff
@@ -0,0 +1,134 @@
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { I_ExchangeWrapper } from "../external/I_ExchangeWrapper.sol";
+import { I_StarkwareContract } from "../external/I_StarkwareContracts.sol";
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "hardhat/console.sol";
+
+contract CurrencyConvertor {
+    using SafeERC20 for IERC20;
+
+    // ============ State Variables ============
+
+    ERC20 immutable USDC_ADDRESS;
+
+    uint256 immutable USDC_ASSET_TYPE;
+
+    I_StarkwareContract public immutable STARKWARE_CONTRACT;
+
+    // ============ Constructor ============
+
+    constructor(
+        address starkwareContractAddress,
+        ERC20 usdcAddress,
+        uint256 usdcAssetType
+    )
+    {
+        STARKWARE_CONTRACT = I_StarkwareContract(starkwareContractAddress);
+        USDC_ADDRESS = usdcAddress;
+        USDC_ASSET_TYPE = usdcAssetType;
+    }
+
+  // ============ Events ============
+
+      event LogConvertedDeposit(
+        address indexed account,
+        address source,
+        address exchangeWrapper,
+        ERC20 tokenFrom,
+        uint256 tokenFromAmount,
+        uint256 tokenToAmount
+    );
+
+  // ============ State-Changing Functions ============
+
+  /**
+    * @notice Make a deposit to the Starkware Layer2 Solution, after converting funds to USDC.
+    *  Funds will be withdrawn from the sender and deposited into the StarkWare Layer2 to the starkKey.
+    * @dev Emits LogConvertedDeposit event.
+    *
+    * @param  tokenFrom        The token to convert from.
+    * @param  tokenFromAmount  The amount of `tokenFrom` tokens to deposit.
+    * @param  exchangeWrapper  The ExchangeWrapper contract to trade with.
+    * @param  starkKey         The starkKey of the L2 account to deposit into.
+    * @param  positionId       The positionId of the L2 account to deposit into.
+    * @param  data             Trade parameters for the ExchangeWrapper.
+    */
+  function deposit(
+    ERC20 tokenFrom,
+    uint256 tokenFromAmount,
+    address exchangeWrapper,
+    uint256 starkKey,
+    uint256 positionId,
+    bytes calldata data
+  )
+    external
+    returns (uint256)
+  {
+
+    // Send fromToken to the ExchangeWrapper.
+    IERC20(tokenFrom).safeTransferFrom(
+      msg.sender,
+      exchangeWrapper,
+      tokenFromAmount
+    );
+
+    address self = address(this);
+
+    // Convert fromToken to toToken on the ExchangeWrapper.
+    I_ExchangeWrapper exchangeWrapperContract = I_ExchangeWrapper(exchangeWrapper);
+
+    uint256 tokenToAmount = exchangeWrapperContract.exchange(
+        msg.sender,
+        self,
+        USDC_ADDRESS,
+        tokenFrom,
+        tokenFromAmount,
+        data
+    );
+
+    // Receive toToken from the ExchangeWrapper.
+    // IERC20(USDC_ADDRESS).safeTransferFrom(
+    //     exchangeWrapper,
+    //     self,
+    //     tokenToAmount
+    // );
```

Comment:
I believe you can't perform this `safeTransferFrom` call because the exchange wrapper contract did not approve the currency convertor contract to pull USDC from it.

You can verify this by logging the following function call in the contract:

`console.log(IERC20(USDC_ADDRESS).allowance(exchangeWrapper, address(this));`

Or the following function call in the tests:

`console.log(await usdc.allowance(zeroExExchangeWrapper.address, currencyConvertor.address));`

I would recommend reading some more documentation online about how allowances work, they can be confusing at first. As a rule of thumb, whenever you are using `safeTransferFrom` and you haven't set an allowance (allowing spender address to move origin addresses ERC20 funds), the transaction will likely fail because of an insufficient allowance.

If you want the currency convertor to transfer funds from the exchange wrapper contract, then you will need to add logic for the exchange wrapper contract to allow the currency convertor contract to spend USDC on its behalf.

If this is what you want to do, I would recommend adding this logic in the exchange wrapper constructor. It would look similar to this:

`IERC20(USDC_ADDRESS).approve(currencyConvertor.address, <amount>)`.

I would recommend passing in the maximum solidity unsigned integer value (we normally use `uint(-1)` in our contracts, but that might not work in Solidity 0.8.0 due to [SafeMath](https://forum.openzeppelin.com/t/using-the-maximum-integer-in-solidity/3000)) for `amount`, since you wouldn't want to manually approve each trade.